File: it-agents/skills/code-architect/references/code-examples/ruby/memento.rb
# frozen_string_literal: true

# Memento — capture and restore an object's state without breaking encapsulation.

# Memento: opaque snapshot. Originator controls what it contains.
class Memento
  def initialize(state) = @state = state
  protected attr_reader :state # only the originator reads it back
end

# Originator creates and restores snapshots of its private state.
class Editor
  def initialize = @content = ""

  def type(text)
    @content += text
    self
  end

  def content = @content

  def save = Memento.new(@content.dup)

  def restore(memento)
    @content = memento.send(:state).dup
    self
  end
end

# Caretaker keeps mementos without inspecting their contents.
class History
  def initialize = @stack = []
  def push(memento) = @stack.push(memento)
  def pop = @stack.pop
end

if __FILE__ == $PROGRAM_NAME
  editor = Editor.new
  history = History.new

  editor.type("hello ")
  history.push(editor.save)
  editor.type("world")

  puts "before undo: #{editor.content.inspect}"
  editor.restore(history.pop)
  puts "after undo:  #{editor.content.inspect}"
end

File: it-agents/skills/code-architect/references/code-examples/php/Memento.php
<?php
// Memento — capture and restore an object's state without breaking encapsulation.
declare(strict_types=1);

// Memento exposes nothing mutable to the outside world.
final class Snapshot
{
    public function __construct(private readonly string $state) {}

    // Only the originator reads the stored state.
    public function state(): string
    {
        return $this->state;
    }
}

// Originator creates and restores from snapshots.
final class Editor
{
    private string $content = '';

    public function write(string $text): void
    {
        $this->content .= $text;
    }

    public function content(): string
    {
        return $this->content;
    }

    public function save(): Snapshot
    {
        return new Snapshot($this->content);
    }

    public function restore(Snapshot $snapshot): void
    {
        $this->content = $snapshot->state();
    }
}

// Caretaker keeps history but never inspects memento internals.
final class History
{
    /** @var Snapshot[] */
    private array $stack = [];

    public function push(Snapshot $snapshot): void { $this->stack[] = $snapshot; }
    public function pop(): ?Snapshot { return array_pop($this->stack); }
}

// Demo
$editor = new Editor();
$history = new History();

$editor->write('Hello');
$history->push($editor->save());
$editor->write(' World');

echo $editor->content() . PHP_EOL; // Hello World
$editor->restore($history->pop());
echo $editor->content() . PHP_EOL; // Hello
